### PR TITLE
Common neural network initializers `nn.initializers`

### LINKED ISSUE #436: [Feature] Add initializers APIs for mlx.nn
Both JAX and PyTorch have initializer classes to initialize weight and bias of a nn module:
https://jax.readthedocs.io/en/latest/jax.nn.initializers.html
https://pytorch.org/docs/stable/nn.init.html

If I am not mistaken, looks like in mlx we currently don't have such initializers, and we need to manually set weight and bias for a mlx.nn object. If we have APIs for some predefined initialization methods (Xavier, Kaiming, etc.) I think it will be easier for users to initialize their neural network modules.

### PR DESCRIPTION
**Pull request by LeonEricsson**
## Proposed changes
closes #436 

This PR implements some common neural network layer initializers as a new module `nn.initializers`. The implementation takes a functional approach, in JAX style as opposed to PyTorch. I tried to find some internal documents on the design philosophy but came up empty handed, in the end I came to the conclusion that this approach is in accordance with the rest of this project. 

The PR is currently in draft as I'd like a review on the first three initializers, `constant`, `normal` and `uniform`, before moving on.

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: python/mlx/nn/initializers.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+def normal(mean: float = 0.0, std: float = 1.0, dtype: mx.Dtype = mx.float32) -> mx.array:
+    r"""Build an initializer that returns an array with random values from a normal distribution.
+
+    Args:
+        dtype (Dtype): The data type of the array.
+        mean (float): Mean of the normal distribution.
+        std (float): Standard deviation of the normal distribution.
+    """
+    def initializer(shape: List[int]) -> mx.array:
```
Comment (gboduljak) [inline python/mlx/nn/initializers.py:24]: Perhaps we could add ``mean`` and ``std`` to ``mx.random.normal``?

Comment (LeonEricsson) [reply python/mlx/nn/initializers.py:24]: yes this would be prefered, not sure if that falls appropriately under this PR though.

**Comment by awni** [comment]
I'm curious about the choice to go with the functional style (e.g. each initializer returns a callable). What motivated that? How are these functions intended to be used?

**Comment by awni** [comment]
PS thanks for the contribution! I do think it could be really useful to have some standard initializers in `nn.initializers`.

**Comment by LeonEricsson** [comment]
> I'm curious about the choice to go with the functional style (e.g. each initializer returns a callable). What motivated that? How are these functions intended to be used?

Hmm, my thought process was that this gives a clean way to set multiple layers using the same initializer. I also wanted something that can work well with Module.apply(). I'm not married to this solution, I also don't build a lot of networks nowadays so my intuition here isn't great.

**Comment by awni** [comment]
Interesting, while you can "mutate" arrays, I don't think we want to encourage that in general as it can get messy to keep track of which references to the same array from a user perspective.

Having something that works with `tree_map` seems like the right call to me so in that sense returning a callable is actually pretty sensible. But to make it work easily we would probably want it to take an array and return another array?

For example with your current implementation, I have to do:

```
my_init = nn.initializer.constant(mx.float16)
tree_map(lambda p: my_init(p.shape), model.parameters())
```

And it would be kind of nice to just do:

```
tree_map(nn.initializer.constant(mx.float16), model.parameters())
```

**Comment by LeonEricsson** [comment]
> And it would be kind of nice to just do:
> 
> ```
> tree_map(nn.initializer.constant(mx.float16), model.parameters())
> ```

yeah I like that.

**Comment by LeonEricsson** [comment]
@awni thoughts now? torch has a bunch of [recommended gain values](https://pytorch.org/docs/stable/nn.init.html#torch.nn.init.calculate_gain) based on the used nonlinearity, but I can't find a source for these?

**Comment by awni** [comment]
@LeonEricsson very nice! We don't have to do it all in this PR. I think what you have is a great start and enough already to merge.

Could you add tests for each initializer? Basic tests are fine, code runs and gives you an array with the expected shape and type. Don't think we need more than that.

**Comment by awni** [comment]
> just remembered I was planning to add an optional key/seed parameter

Oh right yes I meant to comment about that as well!

**Comment by awni** [comment]
> sure np. any docs that need updating outside the file itself?

Yes.

1. Add an initializers section [here](https://github.com/ml-explore/mlx/blob/main/docs/src/python/nn.rst) (like below loss functions).
2. Add a new file with the list of initializers. Follow [losses](https://github.com/ml-explore/mlx/blob/main/docs/src/python/nn/losses.rst) as an example

**Comment by LeonEricsson** [comment]
> 1. Add an initializers section [here](https://github.com/ml-explore/mlx/blob/main/docs/src/python/nn.rst) (like below loss functions).

Do you mean in the Quick Start example or a new subsection after Value and Grad

**Comment by awni** [comment]
I meant under Neural Networks and put it right below Losses

**Comment by LeonEricsson** [comment]
i _think_ the docs are done and it's ready for a merge

**Comment by awni** [comment]
Ok great, I am reviewing now.

**Comment by awni** [comment]
I spent a little time fine-tuning the doc strings a bit. But did not finish. Make sure you pull before making any changes as I pushed to your branch.

File: python/mlx/nn/initializers.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    fan_in = x.shape[0]
+    fan_out = x.shape[1]
+    receptive_field = 1
+
+    if x.ndim > 2:
+        for d in x.shape[2:]:
+            receptive_field *= d
+
+        fan_in = fan_in * receptive_field
+        fan_out = fan_out * receptive_field
```
Comment (awni) [inline python/mlx/nn/initializers.py:122]: We should double check this for the common cases (namely linear and conv). I think for our linear layers we do:

`x @ w.T` so the output dimension would be the first dim of `w`? Also would be good to check for convolutions which have out channels first and in channels last. So maybe the correct thing is:

`fan_in = x.shape[-1]` and `fan_out = x.shape[0]`?

File: python/mlx/nn/__init__.py [hunk 3]
```diff
@@ -1,5 +1,5 @@
 # Copyright © 2023 Apple Inc.
 
-from mlx.nn import losses
+from mlx.nn import initializers, losses
```
Comment (awni) [inline python/mlx/nn/__init__.py:3]: How do you feel about changing the name to `init`? It is much shorter / easier to type :) and PyTorch uses it (we tend to follow PyTorch nn naming conventions).

**Review by awni — CHANGES_REQUESTED** [review]
This is really nice! I left two comments could you take a look and address?

If you have time to add some more examples to the docstrings (like I did for uniform) that would also be great. I can work on it some more tomorrow.

**Review by awni — APPROVED** [review]
I made the change to `nn.init`, hope that's ok 😄 

Thanks for adding this!!

**Comment by mzbac** [comment]
Nice new feature! can't wait to try he_uniform in lora linear :p

**Comment by LeonEricsson** [comment]
> I made the change to `nn.init`, hope that's ok 😄
> 
> Thanks for adding this!!

thanks for patching things up, hectic schedule atm  🙏
